### DISCUSSION TITLE
Improving Performance - Maven Artifact Caching

### DISCUSSION THREAD

**Discussion by goneall** [aquasecurity/trivy#8118] [Q&A]
### Question

I'm running some very large Trivy scans with several hundred Maven POM files being analyzed.

It is taking over 11 hours to complete.

From analysis, almost all of the time is waiting for Maven artifact downloads.

I was wondering if I could configure Trivy to use a local artifact repository which would essentially be a very large cache or mirror of Maven to speed things up.  Something like Sonatype Nexus.

Is it possible to configure Trivy to utilize such a repository?

I did find an [old article]( https://trivy.dev/v0.25.4/docs/references/troubleshooting/#maven-rate-limiting-inconsistent-jar-vulnerability-reporting) on implementing a caching forward proxy to solve a slightly different problem.  This is something I could try if you think it would help.

### Target

Filesystem

### Scanner

License

### Output Format

SPDX

### Mode

Standalone

### Operating System

Ubuntu

**Comment by knqyf263** [comment]
@DmitriyLewen Can you please take a look? You already have many tasks, so anytime you have a minute is fine.

**Reply by DmitriyLewen** [reply]
I will try to check that on this weak.

**Comment by DmitriyLewen** [comment]
Hello @goneall 
I looked at your case.

> I did find an [old article](https://trivy.dev/v0.25.4/docs/references/troubleshooting/#maven-rate-limiting-inconsistent-jar-vulnerability-reporting) on implementing a caching forward proxy to solve a slightly different problem

That instruction was for `jar` files.
No, it doesn't work now because we use trivy-java-db.

About `pom.xml` files:
Trivy doesn't have an env (like `MAVEN_CENTRAL_URL`) to set repositories to look for pom files.
Trivy only checks repositories from `pom.xml` files (so you can temporarily add your local repository to the scanned pom file (but I'm not sure if that's ok for you)).
I think the best solution is to use the local repository from the settings file. We already have #7807 with a request to add this functionality.

**Reply by goneall** [reply]
Thanks @DmitriyLewen - I'll look into your suggestion.  

In looking at the Trivy and trivy-java-db code, it looks like Trivy is already doing quite a bit of caching and network access optimization.

I'll do some further analysis to see if I can come up with some additional optimizations.

**Reply by DmitriyLewen** [reply]
> Trivy is already doing quite a bit of caching 

There is problem with cache for `pom.xml` files:
Trivy doesn't use cache for `fs` mode.

___
Actually i was thinking about this case again last night.
Maybe you are not the only one who faces such a case.
Maybe we should make an exception and add env (e.g. `TRIVY_MAVEN_MIRROR`) with repository to add it first in the list of repositories (because way with  `settings.xml` is also not perfect, since for scanning users will have to change this file (and they can forget to change it back, etc.))
@knqyf263 wdyt?

**Reply by knqyf263** [reply]
@goneall How do you configure mirror repositories with Maven? If it's already configured for Maven, it sounds straightforward that Trivy respects that.

**Reply by goneall** [reply]
We're not running the scans on the same environment as the builds.  This may be a somewhat unique situation - the project is scanning all of the Linux Foundation key projects, so we have to deal with a wide variety of languages and configurations.

From looking at the code - it seems like adding a configuration to the Maven settings.xml should direct the requests to a local repo.  Over the next week or two, I'll setup a couple experiments to see if I can speed it up.  I'll let you know how it goes.

**Reply by goneall** [reply]
Just a short update - I've been doing a few experiments on different scenarios over the last week.  I need to do a bit more analysis before proposing any solutions.

One thing that is likely to help is to enable caching of POM files for file system analysis - but I'll do some some more analysis to see how much this could help.

**Reply by goneall** [reply]
Completed some analysis.

Here's a few conclusions along with some reasoning and data for the conclusions:
- We're impacted by rate limiting on one or more public Maven repositories based on:
  - Running the scan with the offline option runs in 8 minutes whereas running the scan without the option takes 9+ hours
  - Most of the time we're blocked waiting for network IOs
  - The log files shows a consistent 30 seconds between long attempts to resolve POM file locations
- The caching implemented in Trivy does not help - it looks like we're always going out to the internet based on:
  - Running the same scan more than once does not significantly change the time
  - The log files shows the same failures attempting to fetch from remote repositories
  - Based on [this comment](https://github.com/aquasecurity/trivy/discussions/8118#discussioncomment-11661675) - it looks like Trivy doesn't cach in FS mode
- There are several remove repositories involved that may be rate limiting
  - It is difficult to tell exactly which remotes are rate limiting since we aren't currently logging successful fetches from the repo
  - The log messages show several repos with failed fetches, but little or no failed resolves, so one of the alternative remotes are successful
- Using an artifact repository doesn't seem practical for our particular use case
  - The local repository needs to be populated
  - Running a Maven build will populate the local cache / repo, but we don't do builds before we run Trivy - Note: this may be a somewhat unique use case
  - Based on experiments and looking a the code, Trivy will not populate the local repo
  - Mirroring the entire remote repository is possible, but would likely involve a lot of resources - for example, the Maven Central repo is over 45 Terabytes in size

Based on the analysis, implementing a cache for the POM files which persists across runs of Trivy may be the best option.  It looks like much of this is implemented.  If there was some protocol for a "pluggable" cache, I could implement something outside of Trivy.

@DmitriyLewen - how hard would it be to enable the caching for the file system scans?  Is there some way I can help?

Also - feel free to challenge any of my conclusions or provide any alternative solutions.

I have a lot of data from the logs of Trivy and Maven runs with debug enabled I can share if interested including a spreadsheet which analyzes the Trivy debug statements.

**Reply by DmitriyLewen** [reply]
> Running the scan with the offline option runs in 8 minutes whereas running the scan without the option takes 9+ hours

In this case, the report doesn't include most of the artifacts 😞

> how hard would it be to enable the caching for the file system scans?

Caching for `fs` mode doesn't make sense, because we can't create safe key for bucket:
- file can be changed, so we can't use filepath, file info, etc.
- we can use hash of file, but for that we still need to read file (it doesn't make sense)

We can think about adding cache for pom file fetched from remote repository:
But there is still a problem with the modified file.
Only `Sonatype` (and perhaps maven central) forbids modifying artifacts (https://github.com/aquasecurity/trivy-java-db/pull/48#discussion_r1897189289), but we can't be sure that other repositories won't modify the artifact (in which case Trivy will use incorrect/old artifact info).

**Reply by goneall** [reply]
Thanks @DmitriyLewen for the reply.

> We can think about adding cache for pom file fetched from remote repository:
> But there is still a problem with the modified file.
> Only Sonatype (and perhaps maven central) forbids modifying artifacts (https://github.com/aquasecurity/trivy-java-db/pull/48#discussion_r1897189289), but we can't be sure that other repositories won't modify the artifact (in which case Trivy will use incorrect/old artifact info).

From what I can tell from the logs and code, caching the POM files would be sufficient to avoid the rate limiting.  I've always assumed the maven central artifacts to be immutable, so I believe caching POM files from Maven central and Sonatype should work.  I can check with some folks I know at Sonatype to verify.

For the other repos, we could adopt some aging protocol (any artifact older the XX days), but that would complicate the configuration.

Is there any way for me to implement an external caching module for POM files without having to modify the project POM files?  From looking at the debug traces and code, it didn't look possible, but I thought I would check.

**Reply by DmitriyLewen** [reply]
> For the other repos, we could adopt some aging protocol (any artifact older the XX days), but that would complicate the configuration.

This may be unreliable, as changes may happen at any time.
Also people will have to figure out how this caching works, repository specifics, etc. Not all users are ready to deal with this.

But I think we can create a new `TRIVY_EXPERIMENTAL_*` environment (like for RPM archives (https://trivy.dev/latest/docs/coverage/others/rpm/#sbom)) with a list of repositories.
The user will choose the repositories they want to cache (for example, if the user is sure that the artifacts in this repository are immutable).
This will give more options to advanced users and will not affect users who want Trivy out of the box.
@knqyf263 wdyt?

**Reply by goneall** [reply]
> But I think we can create a new TRIVY_EXPERIMENTAL_* environment (like for RPM archives (https://trivy.dev/latest/docs/coverage/others/rpm/#sbom)) with a list of repositories. 

I think this will work for my particular use case

**Reply by goneall** [reply]
Just a quick update - I've been running a couple of experiments.  I've added some logging to narrow down which POM file fetches are slow and experimented with forcing the fetching of the POM files to go through a Nexus cache.

If I limit the fetches to only use Maven Central and go through Nexus, the time goes from around 11 hours to approximately 90 minutes.  I did verify that if I change the following line to go through the Nexus cache, it works and the cache utilized:

https://github.com/aquasecurity/trivy/blob/02ebb4cb89f912e9a62d01dc63523ee93e9d5033/pkg/dependency/parser/java/pom/parse.go#L30

Watching the log files real-time, it looks like the Apache staging repo is rate limiting the responses.  Once I identify the slow repos, I'll see if I can configure them to go through the Nexus cache.  If that speeds things up significantly, I'll do a bit more looking at the code to see if there is an easy way to configure the Nexus cache.

**Reply by goneall** [reply]
I finished my performance experiments and have a solution that I've hacked together on my fork of Trivy: https://github.com/goneall/trivy/blob/usenexus/pkg/dependency/parser/java/pom/parse.go

The biggest performance gain was not repeatedly "trying" to fetch a POM file from a repository which did not respond (either an error or timeout).  I didn't implement this directly in the Trivy code, but redirected the repository requests through the Nexus server which has a feature to respond immediately if a previous attempted fetch of the upstream failed.  This is something that could be implemented directly in Trivy without having to implement the caching / proxy implementation.

The second biggest performance gain was going through the Nexus server which caches the POM file locally.  The gains from this was not nearly as large as the gains from avoiding multiple bad fetches, but it was more than measurable.

My "hack" was to change the hard coded URL for the Maven repo to use the Nexus server.  I also implemented a map of proxy URLs and when fetching from a remote URL matching the key, it fetched it from the Nexus server.  I had to also configure the Nexus server.

If we wanted to update Trivy to such an approach, you can fetch the globally configured proxy server from the Maven settings configuration file (same approach used by Maven).

My recommendation, based on the effort and return, would be to implement a feature which doesn't do multiple bad calls.  The caching would be nice, but it would be more difficult and less impact.

If you want to test this out, the bad repo URL I found in more than one LF projects are:

`https://repository.apache.org/content/repositories/releases/`

and

`https://repository.apache.org/content/repositories/staging/`

One example project is the [opendaylight project](https://github.com/opendaylight).  I think (but I'm not completely sure) the yangtools repo can reproduce the problem.

If I get some time, I may change my hack to a more supportable version which just skips the bad repos as recommended - but I have a few higher priority non Trivy related performance problems to chase down.

**Reply by DmitriyLewen** [reply]
hmm...
I like the idea of ​​"disabling" broken repositories (we can only leave 404 errors).
At first I thought we need a switch (I think `TRIVY_EXPERIMENTAL_*` will be enough for now) so that the user will see that Trivy couldn't access the repository, but then I realized that the user either looks at the logs or not (I mean that if the user doesn't look at the logs - he will miss the error and it doesn't matter how many errors there are - one or for each artifact).

I can think of another case - what if the repository "turns off" for a short time
for example, the scan lasts 30 minutes. After 10 minutes, the repository stops responding for 1-2 minutes.
should we exclude this repository?
I think this is a rare case, so I suggest the following logic:
if the repository did not return a response on the first attempt - skip this repository, otherwise continue trying (thinking that this is a temporary failure)
although this may be redundant.

**Reply by goneall** [reply]
You could vary the policy based on the response from the server - e.g. have a different policy for 502 responses vs. timeouts.

I wish I would have recorded the response code for the Apache calls.

Having some kind of timeout on the exclusion seems reasonable.  This is the approach Nexus takes - they exclude the repo for one day then try again.  A timeout of 1 hour would not impact the performance of my scans.

Another note - in my experiments, there was never any warnings logged.  It looks like the code currently just moves onto the next repository without logging an error.  I'm thinking you only need to report an error if you can't resolve the dependency at all.

**Reply by DmitriyLewen** [reply]
>  I'm thinking you only need to report an error if you can't resolve the dependency at all.

it make sense 👍
